modules/contracts/contracts/adjudicator/mixins/MixinRespondToChallenge.sol

```diff
@@ -34,29 +32,141 @@ contract MixinRespondToChallenge is LibStateChannelApp, LibAppCaller, MChallenge
-            appIdentity.participants,
-            appState
+        // if req.versionNumber > challenge.versionNumber, will be trying
+        // to progress with setState. if req.versionNumber ==
+        // challenge.versionNumber will be trying to progress by using
+        // setStateWithAtion
+        require(
+            stateReq.versionNumber >= challenge.versionNumber,
+            "respondToChallenge was called with outdated state"
         );
 
-        require(
-            turnTaker == keccak256(action).recover(actionSignature),
-            "Action must have been signed by correct turn taker"
+        // if the app state in the `stateReq` is *not* the current
+        // state of the challenge, then it is advancing by several nonces
+        // and you DO NOT apply the action
+
+        // NOTE: this logic indicates that if you need to respond to
+        // a challenge with an action on a much higher nonced state,
+        // you will need to first call `respondToChallenge` with a higher
+        // nonced state, then you will need to call `respondToChallenge`
+        // with the higher nonce state + valid action
+        if (keccak256(stateReq.appState) == challenge.appStateHash) {
+
+          // assert that req.versionNumber == challenge.versionNumber
+          require(
+            stateReq.versionNumber == challenge.versionNumber,
+            "respondToChallenge was called with an incorrect state"
+          );
+
+          // assert the turn taker signed the action
+          require(
+            correctKeySignedTheAction(
+                appIdentity.appDefinition,
+                appIdentity.participants,
+                stateReq,
+                actionReq
+            ),
+            "respondToChallenge called with action signed by incorrect turn taker"
+          );
+
+          require(stateReq.timeout > 0, "Timeout must be greater than 0");
+
+          // This should throw an error if reverts
+          bytes memory newState = LibAppCaller.applyAction(
+              appIdentity.appDefinition,
+              stateReq.appState,
+              actionReq.encodedAction
+          );
+
+          // do not apply the timeout of the challenge update
+          // to the resultant state, instead use the default timeout.
+          // Doing otherwise could violate the signers intention. For
+          // example:
+          // Signer may be fine signing a very small timeout for a favorable
+          // state. ("I made the last move in this state, so I'll win if i
+          // finalized!"). Then counterparty applies an action to it, and now 
+          // signer has very little time to react to this potentially
+          // unfavorable state. ("Now they made the last move and will win!")
+          // instead use the default timeout.
+          uint256 finalizesAt = block.number + appIdentity.defaultTimeout;
+          require(finalizesAt >= appIdentity.defaultTimeout, "uint248 addition overflow");
```

Comment:
Seems like we could use safeMath to implicitly have these overflow checks, but I guess it's fine.

Reply:
im not sure why were not using safe math tbh, it seems like a no brainer. i was just trying to be consistent with the other contracts